qiita_db/portal.py

```diff
@@ -5,20 +5,156 @@
+from qiita_core.exceptions import IncompetentQiitaDeveloperError
 
-# Only make functions available for main portal
-if qiita_config.portal == "QIITA":
-    def get_studies_by_portal(portal):
-        """Returns study id for all Studies belonging to a portal
+
+class Portal(QiitaObject):
+    _table = 'portal_type'
+
+    def __init__(self, portal):
+        self.portal = portal
+        portal_id = convert_to_id(portal, 'portal_type', 'portal')
+        super(Portal, self).__init__(portal_id)
+
+    @classmethod
+    def create(cls, portal, desc):
+        """Creates a new portal on the system
+
+        Parameters
+        ----------
+        portal : str
+            The name of the portal to add
+        desc : str
+            Description of the portal
+
+        Raises
+        ------
+        QiitaDBDuplicateError
+            Portal already exists
+        """
+        if cls.exists(portal):
+            raise QiitaDBDuplicateError("Portal", portal)
+
+        # Add portal and default analyses for all users
+        sql = """DO $do$
+            DECLARE
+                pid bigint;
+                eml varchar;
+                aid bigint;
+            BEGIN
+                INSERT INTO qiita.portal_type (portal, portal_description)
+                VALUES (%s, %s) RETURNING portal_type_id INTO pid;
+
+                FOR eml IN
+                    SELECT email FROM qiita.qiita_user
+                LOOP
+                    INSERT INTO qiita.analysis
+                    (email, name, description, dflt, analysis_status_id)
+                    VALUES (eml, eml || '-dflt', 'dflt', true, 1) RETURNING
+                    analysis_id INTO aid;
+
+                    INSERT INTO qiita.analysis_workflow (analysis_id, step)
+                    VALUES (aid, 2);
+
+                    INSERT INTO qiita.analysis_portal
+                    (analysis_id, portal_type_id)
+                    VALUES (aid, pid);
+                END LOOP;
+            END $do$;"""
+        conn_handler = SQLConnectionHandler()
+        conn_handler.execute(sql, [portal, desc])
+
+        return cls(portal)
+
+    @staticmethod
+    def delete(portal):
+        """Removes a portal from the system
 
         Parameters
         ----------
         portal : str
-           Portal to check studies belong to
+            The name of the portal to add
+
+        Raises
+        ------
+        QiitaDBError
+            Portal has analyses or studies attached to it
+        """
+        conn_handler = SQLConnectionHandler()
+        # Check if attached to any studies
+        portal_id = convert_to_id(portal, 'portal_type', 'portal')
+        sql = """SELECT study_id from qiita.study_portal
+                 WHERE portal_type_id = %s"""
+        studies = conn_handler.execute_fetchall(sql, [portal_id])
+        if studies:
+            raise QiitaDBError(
+                "Studies still attached to portal %s: %s" %
```

Comment:
Right on, if it's not a concern to anyone else, it should be good to
leave as IDs :+1:.


On (Jul-01-15| 8:16), Joshua Shorenstein wrote:
> > +            The name of the portal to add
> > +
> > +        Raises
> > +        ------
> > +        QiitaDBError
> > +            Portal has analyses or studies attached to it
> > +        """
> > +        conn_handler = SQLConnectionHandler()
> > +        # Check if attached to any studies
> > +        portal_id = convert_to_id(portal, 'portal_type', 'portal')
> > +        sql = """SELECT study_id from qiita.study_portal
> > +                 WHERE portal_type_id = %s"""
> > +        studies = conn_handler.execute_fetchall(sql, [portal_id])
> > +        if studies:
> > +            raise QiitaDBError(
> > +                "Studies still attached to portal %s: %s" %
> 
> Aliases can quickly get really long and messy, especially if a long list of studies are already attached to the portal (Think trying to delete EMP or QIITA currently). This error will only be seen by admins on the command line or in the system error logging, so having IDs is probably fine. Will reword the error message though.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiita/pull/1316/files#r33688693